core/libp2p/multi/multiaddress/multiaddress.cpp

```diff
@@ -0,0 +1,233 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <numeric>
+#include <stdexcept>
+
+#include "libp2p/multi/multiaddress.hpp"
+extern "C" {
+#include "libp2p/multi/multiaddress/c-utils/protoutils.h"
+}
+
+namespace {
+  // string representations of protocols
+  constexpr std::string_view kIp4 = "/ip4";
+  constexpr std::string_view kIp6 = "/ip6";
+  constexpr std::string_view kIpfs = "/ipfs";
+  constexpr std::string_view kTcp = "/tcp";
+  constexpr std::string_view kUdp = "/udp";
+  constexpr std::string_view kDccp = "/dccp";
+  constexpr std::string_view kSctp = "/sctp";
+  constexpr std::string_view kUdt = "/udt";
+  constexpr std::string_view kUtp = "/utp";
+  constexpr std::string_view kHttp = "/http";
+  constexpr std::string_view kHttps = "/https";
+  constexpr std::string_view kWs = "/ws";
+  constexpr std::string_view kOnion = "/onion";
+  constexpr std::string_view kWebrtc = "/libp2p-webrtc-star";
+
+  /**
+   * Find beginning of at least one of the provided substrings in the string
+   * @param string to search in
+   * @param substrings to be searched for
+   * @return position of one of the substrings inside of the string or npos, if
+   * none of them is found
+   */
+  size_t findOneOfSubstrings(
+      std::string_view string,
+      std::initializer_list<std::string_view> substrings) {
+    auto pos = std::string_view::npos;
+    std::any_of(std::begin(substrings),
+                std::end(substrings),
+                [&pos, &string](auto substring) {
+                  pos = string.find(substring);
+                  return pos != std::string_view::npos;
+                });
+    return pos;
+  }
+
+  /**
+   * Find all occurrences of the string in other string
+   * @param string to search in
+   * @param substring to be searched for
+   * @return vector with positions of all occurrences of that substring
+   */
+  std::vector<size_t> findSubstringOccurrences(std::string_view string,
+                                               std::string_view substring) {
+    std::vector<size_t> occurrences;
+    auto occurrence = string.find(substring);
+    while (occurrence != std::string_view::npos) {
+      occurrences.push_back(occurrence);
+      occurrence = string.find(substring, occurrence + substring.size());
+    }
+    return occurrences;
+  }
+}  // namespace
+
+namespace libp2p::multi {
+
+  Multiaddress::FactoryResult Multiaddress::createMultiaddress(
+      std::string_view address) {
+    uint8_t *bytes_ptr = nullptr;
+    size_t bytes_size = 0;
+
+    // convert string address to bytes and make sure they represent valid
+    // address
+    auto conversion_error = string_to_bytes(
+        &bytes_ptr, &bytes_size, address.data(), address.size());
```

Comment:
what is the point of creating nullptr bytepointer, and then pass it to some function?

Reply:
This bytepointer is going to be filled the data inside the function. Also, it is the way developers meant to use (and actually used) this function in their original code. But, if you know, how to pass this argument clearer, please write

Reply:
>This bytepointer is going to be filled the data inside the function

but pointer is not allocated